processor/groupbytraceprocessor/processor.go

```diff
@@ -0,0 +1,344 @@
+	// traceID in the map and the spans to the storage
+
+	// calculates the item in the ring that we'll store the trace
+	sp.index = (sp.index + 1) % sp.config.NumTraces
+
+	// see if the ring has an item already
+	idAtRingItem := sp.ids[sp.index]
+	if idAtRingItem != nil {
+		// delete from the map and erase its memory entry
+		delete(sp.idToIndex, hash(idAtRingItem))
+		sp.ids[sp.index] = nil
+
+		// delete from the storage
+		sp.removed <- idAtRingItem
+
+		// TODO: do we want another channel that receives evicted items? record a metric perhaps?
+		sp.logger.Debug("trace evicted", zap.ByteString("traceID", idAtRingItem))
+	}
+
+	// place the traceID in memory
+	sp.ids[sp.index] = traceID
+	sp.idToIndex[hashedTraceID] = sp.index
+
+	// we have the traceID in the memory, place the spans in the storage too
+	if err := sp.addSpans(traceID, trace); err != nil {
+		return fmt.Errorf("couldn't add spans to new trace: %w", err)
+	}
+
+	sp.logger.Debug("scheduled to release trace", zap.Duration("duration", sp.config.WaitDuration))
+
+	time.AfterFunc(sp.config.WaitDuration, func() {
+		sp.cancelLock.RLock()
+		if sp.cancelReleases {
+			sp.logger.Debug("processor is stopping, canceling the release of the trace", zap.ByteString("traceID", traceID))
+			return
+		}
+
+		sp.logger.Debug("releasing trace", zap.ByteString("traceID", traceID))
+		sp.expired <- traceID
+		sp.cancelLock.RUnlock()
+	})
+
+	return nil
+}
+
+func (sp *groupByTraceProcessor) processExpired(traceID []byte) error {
+	sp.logger.Debug("processing expired", zap.ByteString("traceID", traceID))
+	hashed := hash(traceID)
+	indexToRemove, found := sp.idToIndex[hashed]
+
+	if !found {
+		// we likely received multiple batches with spans for the same trace
+		// and released this trace already
+		sp.logger.Debug("skipping the processing of expired trace", zap.ByteString("traceID", traceID))
+		return nil
+	}
+
+	// delete from the map and erase its memory entry
+	delete(sp.idToIndex, hashed)
+	sp.ids[indexToRemove] = nil
+
+	// this might block, but we don't need to wait
+	sp.logger.Debug("marking the trace as released", zap.ByteString("traceID", traceID))
+	go sp.markAsReleased(traceID)
+
+	return nil
+}
+
+func (sp *groupByTraceProcessor) markAsReleased(traceID []byte) error {
+	// #get is a potentially blocking operation
+	trace, err := sp.st.get(traceID)
+	if err != nil {
+		return fmt.Errorf("couldn't retrieve trace %q from the storage: %w", traceID, err)
+	}
+
+	if trace == nil {
+		return fmt.Errorf("the trace %q couldn't be found at the storage", traceID)
+	}
+
+	// signal that the trace is ready to be released
+	sp.logger.Debug("trace marked as released", zap.ByteString("traceID", traceID))
+	sp.released <- trace
+	return nil
+}
+
+func (sp *groupByTraceProcessor) onReleased() {
+	for item := range sp.released {
+		// what should we do with the context?
+		// should we have attempted to store the first context we saw? the last? none?
```

Comment:
You can't use first, last or any other context from the individual call to `ConsumeTraces` because it may have been cancelled by the time you decide to asynchronously export data.

Reply:
Yeah, and whoever set the context might have a different notion of context as we have here. Other places (batch processor?) just discards the context, and I think it might indeed be the right approach here as well.